share/wake/lib/core/list.wake

```diff
@@ -802,3 +875,53 @@ export def groupBy (cmp: a => a => Boolean) (list: List a): List (List a) =
             (elem, head, rest), tail
         _ = (elem, Nil), acc
     foldr combo Nil sorted
+
+# Given a list of pairs, find all right elements such that
+# their left element matches a selection function. This
+# can be used to lookup values by keys if the list
+# represents a dictionary for instance.
+#
+# Parameters:
+#  - `selectFn`: The function that selects which keys to filter by
+#  - `list`: The list of key-value pairs to pull selected values out of
+#
+# Examples:
+# ```
+#   def l = ("Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil)
+#   lookup (_==~"Alice") l = 85, Nil
+#   lookup (_==~"Bob") l = 4, 15, Nil
+#   lookup (_==~"Mal") l = Nil
+#   lookup (\_ True) l = 85, 4, 20, 15, Nil
+#   lookup (\_ False) l = Nil
+# ```
+export def lookup (selectFn: a => Boolean): (list: List (Pair a b)) => List b =
+  def select (Pair key value) = if selectFn key then Some value else None
+  mapPartial select
+
+# TODO: Fix docs
+# Given a list of pairs, edit all right elements such that
+# their left element matches a selection function. This
+# can be used to eidts values in a dictionary for instance.
+#
+# Parameters:
+#  - `selectFn`: The function that selects which keys to edit the values of
+#  - `editFn`: The function used to update values
+#  - `list`: The list of key-value pairs to pull selected values out of
+#
+# Examples:
+# ```
+#   def l = ("Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil)
+#   editLookup (_==~"Alice") (_-5) l = "Alice" :-> 80, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil
+#   editLookup (_==~"Bob") (_-5) l = "Alice" :-> 85, "Bob" :-> -1, "Carol" :-> 20, "Bob" :-> 10, Nil
+#   editLookup (_==~"Mal") (_+1000) l = "Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil
+#   editLookup (\_ True) (_+1) l = "Alice" :-> 86, "Bob" :-> 5, "Carol" :-> 21, "Bob" :-> 16, Nil
+#   editLookup (\_ False) (+1) l = "Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil
+# ```
+export def editLookup (selectFn: a => Boolean) (editFn: b => b): (list: List (Pair a b)) => List (Pair a b) =
+  editFilter (selectFn _.getPairFirst) $ editPairSecond $ editFn
+
+export def editLookupFindFail (selectFn: a => Boolean) (editFn: b => Result b Error): (list: List (Pair a b)) => Result (List (Pair a b)) Error =
+  def editPairSecondFindFail (Pair x y) =
+      require Pass ey = editFn y
+      Pass (Pair x ey)
+  editFilterFindFail (selectFn _.getPairFirst) $ editPairSecondFindFail
```

Comment:
This function succeeds when there are no matching fields. I think it should fail.

Reply:
Hmm right, like if you're looking up a key it should tell you that the key is missing. In fact users might want to handle that separately from the error. I think I'm vaguely happy to have this fail with not matching fields but maybe there should be a `editFilterFindFailWith` that accepts the error for the bad case? That's actually a general problem I think.

Setting the error for the user when there are multiple errors feels kinda wrong to me. There are really two different failure modes here but in general when you compose many of these there will be many failure modes. How do we allow users to handle that?

Also consider accessing `.foo.bar` where there's also a `.bar` field. Maybe we should modify the error message to include `<insert error> when accessing field "field"` or something?